### PR TITLE
 Query optimization for federated SQL query

### LINKED ISSUE #8284: [New feature] Federated SQL query and Query optimization are going to sail out.
Hi  community,

As you know, ShardingSphere has made a lot of efforts on SQL  parser and **provided a great independent SQL parser** to help users parse SQL.

Based on this substantial work, we **plan to do query optimization** to optimize the input SQLs from users and produce an optimized SQL query plan to improve query efficiency. Plus, **the federated SQL query feature** (Like join query from different instances)  is another essential highlight for our next release. : )

We will leverage [Apache Calcite](https://calcite.apache.org), an excellent framework to implement two of the features.  Currently, three main work focus are presented here. 
- [x] The investigation of **RBO** (Rule-Based Optimization) and **CBO**  (Cost-Based optimization) ,i.e.,  Hep planner engine and Volcano planner engine.   (**Mechanism, usage, pros and cons**).
- [x] The investigation of Calcite adaptor, especially `TranslatableTable` API.
- [x] How to transform the parsed result of ShardingSphere to the algebra of Calcite
- [x] Call some of the optimization rules to process the relational expression
- [x] Implement Calcite adaptor to join SQLs from different instances
- [x] Combine sharding tables with Calcite adaptors.

Actually, there are plenty of works to do on this issue. We are **in the investigation phase** now and will seek contributors for this issue later. If you are interested in this one, please give it a watch. 😉 

---
### 10th January 2021 Task Update

Hi, community,

Here is the progress update so far.

### Functions

- [x] The **research** on the source code and API of Apache Calcite
- [x] The **design plan** on Apache ShardingSphere and  Apache Calcite
- [x] The implementation of `CalciteLogicSchemaFactory`, `CalciteLogicSchema`, and `CalciteFilterableTable`
- [x] The Calcite JDBC executor
- [x] The **work embedding with** initialization, router, and executor with ShardingSphere
- [ ] Use ShardingSphere **parser** instead of the Calcite parser to parse SQL (SqlNodeConverter, @guimingyue)  (Q2)
- [ ] Custom **planner** for SQL Optimization(PlannerInitializer, @guimingyue )  (Q2)
- [ ] SQL Optimization for `CalciteFilterableTable` (CalciteExecutionSQLGenerator)  (Q2)
- [ ] The implement for `CalciteRawExecutor` (Especially `ResultSetMetadata`)  (Q2)
- [ ] The **SQLRouter** of ShardingSphere to route SQL to `CalciteExecutor`  (Q2)

### Unit test (Q2)
- [x] CalciteContextFactory
- [x] CalciteLogicSchemaFactory(Doing, #8965)
- [x] CalciteJDBCExecutor(Doing, #8883)
- [ ] CalciteRawExecutor(TODO)
- [ ] CalciteExecutionSQLGenerator(TODO)
- [x] CalciteExecutionContextGenerator

### SQL Federation
- [x] Binding tables (Native)
- [x] Single table and Broadcast table  (Native)
- [x] Sharding table and Broadcast table  (Native)
- [x] Broadcast table and Broadcast table  (Native)
- [x] Single table and Single table + UT
- [x] Single table and Sharding table  + UT(Q2)
- [x] Sharding table and Sharding table  + UT (Q2)

### Scenario
- [x] Sharding
- [ ] Sharding + ReplicaQuery  (Q2)
- [ ] Sharding + Encryption  (Q2)
- [x] Sharding + Encryption + ReplicaQuery  (Q2)

### PR DESCRIPTION
**Pull request by guimingyue**
Add a new module `shardingsphere-infra-optimizer`, and this module is based on Apache Calcite framework. The main packages are as follows:

0. `org.apache.shardingsphere.infra.optimizer.converter`, the converter of converting ShardingSphere ast to Calcite ast, or SqlNode in Calcite world, and the Converter of converting SqlNode to RelNode of Calcite.
1. `org.apache.shardingsphere.infra.optimizer.planner`, the core of the optimizer based on Calcite Planner framework. It contains the planner rules(`org.apache.shardingsphere.infra.optimizer.planner.rule`) and the planner process which is response for converting logical plan to physical plan, also be known as execution plan, please refer to `DefaultPlanner` for this process. 
2. `org.apache.shardingsphere.infra.optimizer.rel`, definition and operation of the `RelNode`.
3. `org.apache.shardingsphere.infra.optimizer.schema`, the bridge of ShardingSphere and Calcite metadata.

For the whole execution process, method `org.apache.shardingsphere.infra.optimize.planner.Compiler#compileQuery` is the entrance of converting ast and optimizing relnode, and this method also return the physical `RelNode`(execution plan).

This module is different from the existing module`shardingsphere-infra-optimize` which can be regraded as an adaptor of calcite, and cannot optimize the join operator which can be pushdown to the sharding database. The new module does not denpend on the `shardingsphere-infra-executor` which is different from `shardingsphere-infra-optimize` too. 

This module has not been finished yet, there still many `TODO` be left in some classes of this module, `org.apache.shardingsphere.infra.optimizer.converter.SqlNodeConverter` for example. 

If you have any problems about this module, let me know.

In the next PR, I will commit the executor of the execution plan, which is based on the volcano execution model, after this PR is merged.

Fixes #8284.

Changes proposed in this pull request:
- Add a new module `shardingsphere-infra-optimizer`, and the optimizer is base on the Apache Calcite framework.

### CODE REVIEW

File: shardingsphere-infra/pom.xml [hunk 1]
```diff
@@ -40,5 +40,6 @@
         <module>shardingsphere-infra-context</module>
         <module>shardingsphere-infra-datetime</module>
         <module>shardingsphere-infra-optimize</module>
+        <module>shardingsphere-infra-optimizer</module>
```
Comment (tristaZero) [inline shardingsphere-infra/pom.xml:43]: What's the difference between `shardingsphere-infra-optimize` and `shardingsphere-infra-optimizer`

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-infra/pom.xml (hunk 1 above)
Comment (guimingyue) [reply shardingsphere-infra/pom.xml:43]: The module `shardingsphere-infra-optimize` can be regarded as an adaptor of calcite, and cannot optimize the join operator which can be pushdown to the sharding database. `shardingsphere-infra-optimizer` module does not depend on the `shardingsphere-infra-executor` which is different from `shardingsphere-infra-optimize` too. The `shardingsphere-infra-executor` will depend on the `shardingsphere-infra-optimizer` for the execution of the physical plan.

File: shardingsphere-infra/shardingsphere-infra-optimizer/src/main/java/org/apache/shardingsphere/infra/optimizer/converter/RelNodeConverter.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    /**
+     * validate calcite SqlNode and convert SqlNode to RelNode.
+     * @param sqlNode SqlNode to be validated and converted
+     * @return relational algebra of sql
+     */
+    public RelNode validateAndConvert(final SqlNode sqlNode) {
+        final RelOptCluster cluster = createRelOptCluster();
+        final SqlToRelConverter sqlToRelConverter = new SqlToRelConverter(null, validator, catalog, cluster,
+                StandardConvertletTable.INSTANCE, converterConfig);
+
```
[the remark this answers is not in the document]
Comment (guimingyue) [reply shardingsphere-infra/shardingsphere-infra-optimizer/src/main/java/org/apache/shardingsphere/infra/optimizer/converter/RelNodeConverter.java]: done

**Comment by tristaZero** [comment]
Hi @guimingyue Do you think it is possible to firstly raise PRs into [SqlNodeConverter](https://github.com/apache/shardingsphere/blob/master/shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/core/convert/SqlNodeConverter.java)

**Comment by guimingyue** [comment]
> Hi @guimingyue Do you think it is possible to firstly raise PRs into [SqlNodeConverter](https://github.com/apache/shardingsphere/blob/master/shardingsphere-infra/shardingsphere-infra-optimize/src/main/java/org/apache/shardingsphere/infra/optimize/core/convert/SqlNodeConverter.java)

It‘s OK. I will try to raise another PR with SqlNodeConverter.

**Comment by guimingyue** [comment]
@tristaZero I have raised a pr to merge SqlNodeConverter。
There are still many TODO for both ss ast and converter tool, I think the list blow ought to be given priority.
- join type is missing.
- having clause of group by is not supported.
- expression has not been pared.

**Comment by tristaZero** [comment]
@guimingyue Cool, I will give it a look.
Could you look at the latest master branch? I suppose 1 and 2 are supported now, and in terms of 3, could you give more clarification?

```
1. join type is missing.
2. having clause of group by is not supported.
3. expression has not been pared.
```

**Comment by guimingyue** [comment]
> @guimingyue Cool, I will give it a look.
> Could you look at the latest master branch? I suppose 1 and 2 are supported now, and in terms of 3, could you give more clarification?
> 
> ```
> 1. join type is missing.
> 2. having clause of group by is not supported.
> 3. expression has not been pared.
> ```

I have tested in the latest branch, the result follows.
- joinType of JoinTableSegment is still null for inner join and left outer join according testcase of `org.apache.shardingsphere.infra.optimize.core.convert.SqlNodeConverterTest`  in my PR.
- having clause is supported now, but the inner expression of AggregationProjectionSegment is String, take this sql  text as an example, `select order_id, user_id from t_order where order_id = 10 group by order_id having sum(order_id) > 0`, the type of expr field of `HavingSegment` is `BinaryOperationExpression` and the left parameter type of `BinaryOperationExpression` object is `AggregationProjectionSegment`. As a result, the type of `innerExpression` field is String whose value is `(order_id)`, which should be parsed as `ColumnSegment` object.
- the field text of `ExpressionProjectionSegment` is a Stirng type, which cannot be converted to a AST expression, for example, in the select list of this sql text `select 10 + 30, order_id + 10, order_id, user_id from t_order`, the expression of `10+30`  will be parsed as an `ExpressionProjectionSegment` object.

**Pull request closed without merging**